### ISSUE TITLE
Core Lint failure in LoadIface

### ISSUE THREAD

**Issue by monoidal** [tweag/ghc#420]
Currently, the branch has -O0 in LoadIface, because the build would fail.

I don't know what's going on yet.

**Comment by monoidal** [comment]
I believe this is another instance of #405. I manually reduced the offending Core to

```
case[1] ... of dt_XnO7
{ UDFM ipv_sHsm ipv_sHsn ->
       join {
         $j_sIJZ x = (dt_XnO7, x)
       } in
       joinrec {
         go_XzxE ds_azgk
           = case[Omega] ds_azgk of {
               C1 -> jump $j_sIJZ ds_azgk;
               C2 _ -> jump go_XzxE (f ds_azgk)
             };
       } in jump go_XzxE something
}
````

where we don't consider dt_XnO7 to be linear. There might be something more to this, since when I changed joinrecs to return the bottom environment the bug persisted, but let's fix #405 first.

**Comment by monoidal** [comment]
Closing as a duplicate of #405.
